### PR TITLE
Add additional ignore_headers audit configuration setting

### LINKED ISSUE #3875: Redact custom headers from audit logging
**Is your feature request related to a problem?**

Right now you can specify a custom `jwt_header` in `config.yml` when dictating JWT settings. This allows you to replace the default `Authorization` header with an alternative header such as `My_header`. You then use this custom header to pass your JWT in your requests so instead of `Authorization: Bearer 1234567` you provide `My_header: Bearer 123456`.

Another feature offered by the security plugin is audit logging. Audit logging also has a setting which allows you to redact or remove sensitive Transport and REST headers. One of the sensitive REST headers is the `Authorization` header that is associated with JWT use. Currently this setting is not able to account for custom headers provided in the JWT configuration. The audit log then shows the JWT used for requests because it logs the request headers not matching `Authorization`.

Because of this, it is NOT RECOMMENDED to use both custom headers and audit logging.

Changes should be made to support use of audit logs and custom headers. 

**What alternatives have you considered?**
Leaving as is and filing a documentation issue (already done).

### PR DESCRIPTION
**Pull request by scrawfor99**
### Description
This change adds a new setting to the audit.yml configuration allowing users to specify REST headers they wish to be ignored from audit logging. 

### Issues Resolved

- Resolves #3875 

### Testing
Added a couple new test cases to check the setting

### Check List
- [x] New functionality includes testing
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/security/auditlog/impl/AuditMessage.java [hunk 1]
```diff
@@ -356,12 +356,15 @@ public void addRestParams(Map<String, String> params) {
...
 
-    public void addRestHeaders(Map<String, List<String>> headers, boolean excludeSensitiveHeaders) {
+    public void addRestHeaders(Map<String, List<String>> headers, boolean excludeSensitiveHeaders, AuditConfig.Filter filter) {
         if (headers != null && !headers.isEmpty()) {
             final Map<String, List<String>> headersClone = new HashMap<>(headers);
             if (excludeSensitiveHeaders) {
                 headersClone.keySet().removeIf(AUTHORIZATION_HEADER);
             }
+            if (filter != null) {
+                headersClone.entrySet().removeIf(entry -> filter.isHeaderDisabled(entry.getKey()));
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/auditlog/impl/AuditMessage.java:366]:
```suggestion
                headersClone.entrySet().removeIf(entry -> filter.shouldExcludeHeader(entry.getKey()));
```

**Review by cwperks — APPROVED** [review]
LGTM! I left one small suggestion around naming the method on the filter.

File: src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java [hunk 2]
```diff
@@ -1360,7 +1360,15 @@ public List<Setting<?>> getSettings() {
...
-            ); // not filtered here
+            );
+            settings.add(
+                Setting.listSetting(
+                    ConfigConstants.SECURITY_AUDIT_IGNORE_HEADERS,
+                    Collections.emptyList(),
+                    Function.identity(),
+                    Property.NodeScope
+                )
+            );// not filtered here
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java:1371]: nit: this comment should probly be on line 1363

File: src/main/java/org/opensearch/security/compliance/ComplianceConfig.java [hunk 3]
```diff
@@ -235,6 +235,7 @@ public static ComplianceConfig from(Map<String, Object> properties, @JacksonInje
         final Set<String> ignoredComplianceUsersForRead = ImmutableSet.copyOf(
             getOrDefault(properties, "read_ignore_users", AuditConfig.DEFAULT_IGNORED_USERS)
         );
+
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/compliance/ComplianceConfig.java:238]: nit: extra line. (in couple other files as well)

**Review by DarshitChanpura — APPROVED** [review]
LGTM. Left few nits

**Comment by DarshitChanpura** [comment]
There is a compilation error:
```
/home/runner/work/security/security/src/test/java/org/opensearch/security/auditlog/impl/AuditMessageTest.java:79: error: cannot find symbol
        when(auditConfig.getFilter().isHeaderDisabled("test-header")).thenReturn(false);
```

File: src/main/java/org/opensearch/security/compliance/ComplianceConfig.java (hunk 3 above)
Comment (scrawfor99) [reply src/main/java/org/opensearch/security/compliance/ComplianceConfig.java:238]: Thanks DC, missed the test renames. Added the newlines and fixed the old comment I bumped.

**Review by DarshitChanpura — DISMISSED** [review]

File: src/main/java/org/opensearch/security/compliance/ComplianceConfig.java (hunk 3 above)
Comment (peternied) [reply src/main/java/org/opensearch/security/compliance/ComplianceConfig.java:238]: Missing a push? I still see these

Comment (scrawfor99) [inline src/main/java/org/opensearch/security/compliance/ComplianceConfig.java:238]:
```suggestion
```

File: src/test/java/org/opensearch/security/auditlog/config/AuditConfigSerializeTest.java [hunk 4]
```diff
@@ -370,6 +377,7 @@ public void testCustomSettings() throws IOException {
     private boolean compareJson(final String json1, final String json2) throws JsonProcessingException {
         ObjectNode objectNode1 = objectMapper.readValue(json1, ObjectNode.class);
         ObjectNode objectNode2 = objectMapper.readValue(json2, ObjectNode.class);
+
```
Comment (scrawfor99) [inline src/test/java/org/opensearch/security/auditlog/config/AuditConfigSerializeTest.java:380]:
```suggestion
```

File: config/config.yml [hunk 5]
```diff
@@ -84,6 +84,7 @@ config:
         ###### and here https://tools.ietf.org/html/rfc7239
         ###### and https://tomcat.apache.org/tomcat-8.0-doc/config/valve.html#Remote_IP_Valve
     authc:
+
```
Comment (cwperks) [inline config/config.yml:87]: nit: remove this new line

Comment (scrawfor99) [inline config/config.yml:87]:
```suggestion
```

**Review by cwperks — APPROVED** [review]

**Review by peternied — APPROVED** [review]

**Comment by peternied** [comment]
@scrawfor99 Do you have a link to the documentation update?

**Pull request merged**